airbyte-cdk/python/airbyte_cdk/sources/availability_strategy.py

```diff
@@ -0,0 +1,97 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from abc import ABC, abstractmethod
+from requests import HTTPError
+from typing import Dict, List, Optional, Text, Tuple
+
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from airbyte_cdk.sources.streams import Stream
+
+
+class AvailabilityStrategy(ABC):
+    """
+    Abstract base class for checking stream availability
+    """
+
+    @abstractmethod
+    def check_availability(self, stream: Stream) -> Tuple[bool, any]:
+        """
+        Checks stream availability.
+
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+
+
+class HTTPAvailabilityStrategy(AvailabilityStrategy):
+    def check_availability(self, stream: Stream) -> Tuple[bool, str]:
+        """
+        Check stream availability by attempting to read the first record of the
+        stream.
+
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+        try:
+            # Some streams need a stream slice to read records (e.g. if they have a SubstreamSlicer)
+            stream_slice = self._get_stream_slice(stream)
+            records = stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream_slice)
+            next(records)
+        except HTTPError as error:
+            return False, repr(error)
```

Comment:
Move exception handling to separate method so that instances of this class don't need to worry about the try: part
see: [github](https://github.com/airbytehq/airbyte/pull/19978/files#diff-895c7e2fc3c7a29570263791b1d031ceb5309127e9e42609ddd9ed425f548e47R241)